src/Api/PubnubApi/EndPoint/PubSub/SignalOperation.cs

```diff
@@ -0,0 +1,162 @@
+                }
+                if (!ChannelInternetStatus.ContainsKey(instance.InstanceId))
+                {
+                    ChannelInternetStatus.GetOrAdd(instance.InstanceId, new ConcurrentDictionary<string, bool>());
+                }
+                if (!ChannelGroupInternetStatus.ContainsKey(instance.InstanceId))
+                {
+                    ChannelGroupInternetStatus.GetOrAdd(instance.InstanceId, new ConcurrentDictionary<string, bool>());
+                }
+            }
+        }
+
+        public SignalOperation Message(object message)
+        {
+            this.msg = message;
+            return this;
+        }
+
+        public SignalOperation Channel(string channelName)
+        {
+            this.channelName = channelName;
+            return this;
+        }
+
+        public SignalOperation QueryParam(Dictionary<string, object> customQueryParam)
+        {
+            this.queryParam = customQueryParam;
+            return this;
+        }
+
+        public void Execute(PNCallback<PNPublishResult> callback)
+        {
+#if NETFX_CORE || WINDOWS_UWP || UAP || NETSTANDARD10 || NETSTANDARD11 || NETSTANDARD12
+            Task.Factory.StartNew(() =>
+            {
+                this.savedCallback = callback;
+                Signal(this.channelName, this.msg, null, this.queryParam, callback);
+            }, CancellationToken.None, TaskCreationOptions.PreferFairness, TaskScheduler.Default).ConfigureAwait(false);
+#else
+            new Thread(() =>
+            {
+                this.savedCallback = callback;
+                Signal(this.channelName, this.msg, null, this.queryParam, callback);
+            })
+            { IsBackground = true }.Start();
+#endif
+        }
+
+        internal void Retry()
+        {
+#if NETFX_CORE || WINDOWS_UWP || UAP || NETSTANDARD10 || NETSTANDARD11 || NETSTANDARD12
+            Task.Factory.StartNew(() =>
+            {
+                Signal(this.channelName, this.msg, null, this.queryParam, savedCallback);
+            }, CancellationToken.None, TaskCreationOptions.PreferFairness, TaskScheduler.Default).ConfigureAwait(false);
+#else
+            new Thread(() =>
+            {
+                Signal(this.channelName, this.msg, null, this.queryParam, savedCallback);
+            })
+            { IsBackground = true }.Start();
+#endif
+        }
+
+        private void Signal(string channel, object message, Dictionary<string, object> metaData, Dictionary<string, object> externalQueryParam, PNCallback<PNPublishResult> callback)
+        {
+            if (string.IsNullOrEmpty(channel) || string.IsNullOrEmpty(channel.Trim()) || message == null)
+            {
+                throw new ArgumentException("Missing Channel or Message");
+            }
+
+            if (string.IsNullOrEmpty(config.PublishKey) || string.IsNullOrEmpty(config.PublishKey.Trim()) || config.PublishKey.Length <= 0)
```

Comment:
Should we also check for subscribe key?

Reply:
added check for SubscribeKey also. will merge to PR in few mins.